src/shared/install.c

```diff
@@ -1724,19 +1724,74 @@ static int install_info_symlink_alias(
         assert(config_path);
 
         STRV_FOREACH(s, i->aliases) {
+
                 _cleanup_free_ char *alias_path = NULL, *dst = NULL;
 
                 q = install_full_printf(i, *s, &dst);
                 if (q < 0)
                         return q;
 
+                if (unit_name_is_valid(i->name, UNIT_NAME_TEMPLATE) &&
+                    !unit_name_is_valid(dst, UNIT_NAME_TEMPLATE) &&
+                    !i->default_instance) {
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Aliases for templates without DefaultInstance must be templates.");
+                }
+
+                if (unit_name_is_valid(i->name, UNIT_NAME_INSTANCE | UNIT_NAME_TEMPLATE) &&
+                    !unit_name_is_valid(dst, UNIT_NAME_TEMPLATE | UNIT_NAME_INSTANCE)) {
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Aliases for template instances must be a templates or template instances containing the instance name.");
+                }
+
+                if (unit_name_is_valid(i->name, UNIT_NAME_INSTANCE | UNIT_NAME_TEMPLATE) &&
+                    unit_name_is_valid(dst, UNIT_NAME_INSTANCE)) {
+                        _cleanup_free_ char *src_inst = NULL, *dst_inst = NULL;
+
+                        q = unit_name_to_instance(i->name, &src_inst);
+                        if (q < 0)
+                                return q;
+
+                        q = unit_name_to_instance(dst, &dst_inst);
+                        if (q < 0)
+                                return q;
+                        if (!strstr(dst_inst, src_inst?src_inst:i->default_instance))
```

Comment:
strstr() is sloppy...  also, please whitespace before and after the ? and :...

Reply:
This is reworked in a later commit.